File: GenerateAst.cs
namespace GenerateAst
{
    public class GenerateAst
    {
        public static void Generate(string[] args)
        {
            if (args.Length != 1)
            {
                Console.WriteLine("Usage GenerateAst <output_directory>");
                Environment.Exit(1);
            }
            string output = args[0];
            DefineAst(output, "Expr", [
                "Binary  : Expr Left, Token Op, Expr Right",
                "Grouping: Expr Expression",
                "Literal : object? Value",
                "Unary   : Token Op, Expr Right",
                "Variable: Token name"
            ]);
            DefineAst(output, "Stmt", [
                "Expression : Expr expression",
                "Print      : Expr expression",
                "Var        : Token name, Expr? initializer"
            ]);
        }

        static void DefineAst(string output, string basename, List<string> types)
        {
            string filename = Path.Join(output, basename + ".cs");
            using (StreamWriter outputFile = new StreamWriter(filename, false))
            {
                outputFile.WriteLine("using interpreter;");
                outputFile.WriteLine("namespace Ast {");
                outputFile.WriteLine($"\tpublic abstract class {basename} {{");
                DefineVisitor(outputFile, basename, types);
                outputFile.WriteLine("\t\tpublic abstract T Accept<T>(IVisitor<T> visitor);");
                foreach (string type in types)
                {
                    string className = type.Split(':')[0].Trim();
                    string fields = type.Split(':')[1].Trim();
                    DefineType(outputFile, basename, className, fields);
                }
                outputFile.WriteLine("\t}");
                outputFile.WriteLine("}");
            };
        }

        static void DefineVisitor(StreamWriter writer, string basename, List<string> types)
        {
            writer.WriteLine($"\t\tpublic interface IVisitor<T> {{");
            foreach (string type in types)
            {
                string name = type.Split(':')[0].Trim();
                writer.WriteLine($"\t\t\tpublic T Visit({name} {basename.ToLower()});");
            }
            writer.WriteLine("\t\t}");
        }

        static void DefineType(StreamWriter writer, string basename, string className, string fields)
        {
            writer.WriteLine($"\t\tpublic class {className}({fields}) : {basename} {{");
            string[] fieldList = fields.Split(", ");
            foreach (string field in fieldList)
            {
                string[] typename = field.Split(' ');
                writer.WriteLine($"\t\t\tpublic {typename[0]} {typename[1]} = {typename[1]};");
            }
            writer.WriteLine("\t\t\tpublic override T Accept<T>(IVisitor<T> visitor) {");
            writer.WriteLine("\t\t\t\treturn visitor.Visit(this);");
            writer.WriteLine("\t\t\t}");
            writer.WriteLine("\t\t}");
        }
    }
}

File: GenerateAst.py
from io import TextIOWrapper
import sys
from pathlib import Path

def define_visitor(fp: TextIOWrapper, basename: str, types: list[str]):
    fp.writelines([
        f"\t\tpublic interface IVisitor<T> {{\n"
    ])
    for type in types:
        name = type.split(":")[0].strip()
        fp.writelines([
            f"\t\t\tpublic T Visit({name} {basename.lower()});\n"
        ])
    fp.writelines(["\t\t}\n"])

def define_type(fp: TextIOWrapper, basename: str, class_name: str, fields: str):
    fp.writelines([f"\t\tpublic class {class_name}{'(' + fields + ')' if len(fields) else ''} : {basename} {{\n"])
    if len(fields):
        for field in fields.split(", "):
            typename = field.split(" ")
            fp.writelines([f"\t\t\tpublic {typename[0]} {typename[1]} = {typename[1]};\n"])
    fp.writelines([
        "\t\t\tpublic override T Accept<T>(IVisitor<T> visitor) {\n",
        "\t\t\t\treturn visitor.Visit(this);\n",
        "\t\t\t}\n",
        "\t\t}\n"
    ])

def define_ast(output: Path, basename: str, types: list[str]):
    filename = output.joinpath(basename + ".cs")
    with open(filename, "w") as fp:
        fp.writelines([
            "using interpreter;\n",
            "namespace Ast {\n",
            f"\tpublic abstract class {basename} {{\n",
        ])
        define_visitor(fp, basename, types)
        fp.write("\t\tpublic abstract T Accept<T>(IVisitor<T> visitor);\n")
        for type in types:
            class_name = type.split(":")[0].strip()
            fields = type.split(":")[1].strip()
            define_type(fp, basename, class_name, fields)
        fp.writelines([
            "\t}\n",
            "}\n"
        ])

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage {sys.argv[0]} <output_directory>")
        sys.exit(1)
    dir = Path(sys.argv[1])
    define_ast(dir, "Expr", [
        "Assign  : Token name, Expr value",
        "Binary  : Expr Left, Token Op, Expr Right",
        "Logical : Expr left, Token op, Expr right",
        "Grouping: Expr Expression",
        "Literal : object? Value",
        "Unary   : Token Op, Expr Right",
        "Variable: Token name"
    ])
    define_ast(dir, "Stmt", [
        "Block      : List<Stmt> statements",
        "Expression : Expr expression",
        "If         : Expr condition, Stmt then_branch, Stmt? else_branch",
        "Print      : Expr expression",
        "Var        : Token name, Expr? initializer",
        "While      : Expr condition, Stmt body, Expr? end_of_loop, Stmt? final",
        "Break      :",
        "Continue   :",
    ])